From: Ryan VanderMeulen <rvandermeulen@mozilla.com>
Date: Mon, 05 Oct 2026 09:49:04 -0400
Subject: Bug 2068902 - Use __builtin_prefetch for absl::PrefetchToLocalCacheForWrite on x86-64.

The prefetchw fallbacks assume CPUs lacking PRFCHW treat the opcode as a no-op. 64-bit NetBurst CPUs
(Pentium 4 / Xeon family 15) raise #UD instead. Let the compiler pick the instruction: prefetcht0 by
default, prefetchw when the target guarantees it.
---
 abseil-cpp/absl/base/prefetch.h | 20 +++++---------------
 1 file changed, 5 insertions(+), 15 deletions(-)

diff --git a/abseil-cpp/absl/base/prefetch.h b/abseil-cpp/absl/base/prefetch.h
index 482cde3a0166..96ee28d3e6a3 100644
--- a/abseil-cpp/absl/base/prefetch.h
+++ b/abseil-cpp/absl/base/prefetch.h
@@ -117,7 +117,9 @@ void PrefetchToLocalCacheNta(const void* addr);
 // This function is similar to `PrefetchToLocalCache()` except that it
 // prefetches cachelines with an 'intent to modify' This typically includes
 // invalidating cache entries for this address in all other cache tiers, and an
-// exclusive access intent.
+// exclusive access intent. On targets where the compiler cannot assume a
+// write-prefetch instruction (e.g. x86-64 without PRFCHW), this is equivalent
+// to `PrefetchToLocalCache()`.
 //
 // Incorrect or gratuitous use of this function can degrade performance. As this
 // function can invalidate cached cachelines on other caches and computer cores,
@@ -154,15 +156,7 @@ ABSL_ATTRIBUTE_ALWAYS_INLINE inline void PrefetchToLocalCacheNta(
 
 ABSL_ATTRIBUTE_ALWAYS_INLINE inline void PrefetchToLocalCacheForWrite(
     const void* addr) {
-  // [x86] gcc/clang don't generate PREFETCHW for __builtin_prefetch(.., 1)
-  // unless -march=broadwell or newer; this is not generally the default, so we
-  // manually emit prefetchw. PREFETCHW is recognized as a no-op on older Intel
-  // processors and has been present on AMD processors since the K6-2.
-#if defined(__x86_64__) && !defined(__PRFCHW__)
-  asm("prefetchw %0" : : "m"(*reinterpret_cast<const char*>(addr)));
-#else
   __builtin_prefetch(addr, 1, 3);
-#endif
 }
 
 #elif defined(ABSL_INTERNAL_HAVE_SSE)
@@ -183,12 +177,8 @@ ABSL_ATTRIBUTE_ALWAYS_INLINE inline void PrefetchToLocalCacheForWrite(
     const void* addr) {
 #if defined(_MM_HINT_ET0)
   _mm_prefetch(reinterpret_cast<const char*>(addr), _MM_HINT_ET0);
-#elif !defined(_MSC_VER) && defined(__x86_64__)
-  // _MM_HINT_ET0 is not universally supported. As we commented further
-  // up, PREFETCHW is recognized as a no-op on older Intel processors
-  // and has been present on AMD processors since the K6-2. We have this
-  // disabled for MSVC compilers as this miscompiles on older MSVC compilers.
-  asm("prefetchw %0" : : "m"(*reinterpret_cast<const char*>(addr)));
+#else
+  _mm_prefetch(reinterpret_cast<const char*>(addr), _MM_HINT_T0);
 #endif
 }
 
